pkgs/by-name/in/install-grub-ng/src/grub.rs

```diff
@@ -0,0 +1,305 @@
+
+			if !status.success() {
+				bail!(
+					"Failed to retrieve subvolume info for {}",
+					fs.mount.display()
+				);
+			}
+
+			let mut ids = id_info.lines().filter_map(|line| {
+				if let Ok(l) = line {
+					l.trim()
+						.strip_prefix("Subvolume ID:")
+						.map(|s| s.trim().to_owned())
+				} else {
+					None
+				}
+			});
+
+			let Some(id) = ids.next() else {
+				return Ok(path);
+			};
+
+			if ids.next().is_some() {
+				bail!(
+					"Btrfs subvol name for {} listed multiple times in mount",
+					fs.mount.display()
+				);
+			}
+
+			id
+		};
+
+		let prefix = {
+			let Output {
+				status,
+				stdout: path_info,
+				..
+			} = Command::new(btrfs)
+				.arg("subvol")
+				.arg("list")
+				.arg(&fs.mount)
+				.output()
+				.context("Failed to execute btrfs subvol list")?;
+
+			if !status.success() {
+				bail!(
+					"Failed to find {} subvolume info from btrfs",
+					fs.mount.display()
+				);
+			}
+
+			let mut paths = path_info.lines().filter_map(|line| {
+				let Ok(l) = line else {
+					return None;
+				};
+
+				let mut split = l.split(' ');
+				if split.nth(1) != Some(&subvol_id) {
+					return None;
+				}
+
+				split.skip_while(|&p| p != "path").nth(1).map(String::from)
+			});
+
+			let Some(path) = paths.next() else {
+				bail!("Could not find a subvolume ID for {}", fs.mount.display())
+			};
+
+			if paths.next().is_some() {
+				bail!(
+					"Btrfs returned multiple paths for a single subvolume id, mountpoint {}",
+					fs.mount.display()
+				);
+			}
+
+			path
+		};
+
+		let mut new = Path::new("/").join(prefix);
+		new.push(path);
+		Ok(new)
+	}
+}
+
+#[derive(Clone, Debug, Default)]
+struct Fs {
+	device: PathBuf,
+	fs_type: String,
+	mount: PathBuf,
+}
+impl Fs {
+	fn new(dir: &Path, store_dir: &Path) -> Result<Self> {
+		let mut best = Self::default();
+		let mount_info = BufReader::new(fs::File::open("/proc/self/mountinfo")?);
+
+		for line in mount_info.lines() {
+			let line = line?;
+
+			let mut fields = line.split(' ');
+			let Some(mount_point) = fields.nth(4).map(Path::new) else {
+				bail!("Mount point not found in mountinfo entry: {line}")
+			};
+			// TODO: This is completely unused in the Perl version. Why is it here?
+			let Some(_mount_options) = fields.next().map(|s| s.split(',')) else {
+				bail!("Mount options not found in mountinfo entry: {line}")
+			};
+
+			// Skip the optional fields.
+			let mut fields = fields.skip_while(|&field| field != "-").skip(1);
+
+			let Some(fs_type) = fields.next() else {
+				bail!("Filesystem type not found in mountinfo entry: {line}")
+			};
+			let Some(device) = fields.next() else {
+				bail!("Device not found in mountinfo entry: {line}")
+			};
+			let Some(mut super_options) = fields.next().map(|s| s.split(',')) else {
+				bail!("Super options not found in mountinfo entry: {line}")
+			};
+
+			// Skip the bind-mount for the Nix store.
+			if mount_point == store_dir && super_options.any(|s| s == "rw") {
```

Comment:
I believe it should be checking `mount_options` rather than `super_options` here, as the latter is *always* going to contain `rw`, for both the bind mount and non-bind mount, at least on the systems I've checked.

That would also explain why `mount_options` appears to be unused in the Perl version: they accidentally used the wrong variable here. Although I guess that never mattered enough to cause a bug in practice, or at least not one that anyone tracked down...

Reply:
Ah, that makes *so* much sense actually. I was confused by this as well xD